archunit/src/test/java/com/tngtech/archunit/library/ArchitecturesTest.java

```diff
@@ -146,6 +154,66 @@ public void gathers_all_layer_violations() {
                         "rule with ignore specified as predicates"));
     }
 
+    @Test
+    public void gathers_all_onion_architecture_violations() {
+        OnionArchitecture architecture = onionArchitecture()
+                .domainModel(absolute("onionarchitecture.domain.model"))
+                .domainService(absolute("onionarchitecture.domain.service"))
+                .application(absolute("onionarchitecture.application"))
+                .adapter("cli", absolute("onionarchitecture.adapter.cli"))
+                .adapter("persistence", absolute("onionarchitecture.adapter.persistence"))
+                .adapter("rest", absolute("onionarchitecture.adapter.rest"));
+        JavaClasses classes = new ClassFileImporter().importPackages(getClass().getPackage().getName() + ".testclasses.onionarchitecture");
+
+        EvaluationResult result = architecture.evaluate(classes);
+
+        ImmutableSet<String> expectedRegexes = ImmutableSet.of(
```

Comment:
The naming seems pretty driven by `layeredArchitecture()` :wink: (since they are all called `..LayerClass`)
I would try to keep the language completely independent of `layers`, because

* in an onion architecture shouldn't it be called "ring"? :wink: 
* `layeredArchitecture()` is implementation detail from my point of view
* the whole reason for the `onionArchitecture()` was to better express the concepts, even though you could write the rules with `layeredArchitecture()` (as you've demonstrated within the implementation :wink:). So I would ban all "layer" talk from the API, description and tests and just talk about "domain model", "application" and so on.

Reply:
True, I would love to use an appropriate name as you suggest. But I always find `layer` in any article about onion architectures. So I assume that it is the most appropriate name. It might be confusing if you compare the concept with `LayeredArchitecture`. But on the other hand, `layer` seems to be the standard name. So I prefer to leave it that way.

Reply:
@codecholeric I added some integration tests as well. Thanks for the hint that those exist in the project since they revealed a minor flaw in the layer definitions.
Unfortunately, It was not easily possible to add the onion architecture integration test to the example project. I therefore put it into a separate package to avoid breaking/needing to modify the existing tests. Is that how you expect it? I think there should not be a single project that deals with all test cases as it will get hard to add new features. Moreover, some architectures might contradict each other (e.g. 3-tier vs. onion architecture).

Reply:
Yeah, I guess this project has just grown and it would be better to split parts of the existing example as well. Maybe I'll do that at some point, since the "shopping" example is kind of odd in there, too. 
I would probably choose a naming like "example.subcase" instead of "subcase.example" if I had the chance, but for now I think it's best how you added it :+1:

Reply:
What I wonder is, if the unit test and the integration test really test anything different? I.e. is it necessary to assert all the violation messages here, if we have the same classes again in the integration test and assert the messages there?

Reply:
The main benefit is not that it does another test case but rather shows how to use the new API for JUnit4, JUnit5, and without any JUnit support. But indeed, I thought about that as well.

Reply:
Actually, we could also add a new "more-or-less real world example" instead of those dummy classes that shows the architecture in a better way. But then the tests might get really odd as there are so many error cases to check for if you do not want to miss any layer interaction. But we can have a chat on that. In particular, whether it is worth the effort building a nice example.
Looking at the time, I assume you are heading to Austria right now :wink: and I will head over there tomorrow morning. So maybe we find some time there.

Reply:
I'm actually on holiday :wink: (as you can clearly see as I have time to work on ArchUnit :stuck_out_tongue:)
What I meant is, you can maybe drop the unit test for the specific violation patterns and keep the example. It is true though, that the examples are usually a little closer to "real world", i.e. you would maybe not violate all constraints in every class.
So I would either make the example closer to the "real world" (even though bodies of methods are just comment stubs) or drop the unit test. (the example + integration test I would keep in any case in some way, just a redundant assertion of the same case is superfluous in my opinion)

Reply:
As a follow up to this: I would add a more realistic example, like the shopping example with an onion architecture. It does not need to have every possible violation, just a couple, but those in a somewhat reasonable way (not just writing down all possible accesses in all classes). Then this would still make sense as a unit test, testing all accesses in detail and there would be a more realistic example for users to see, that can also serve as an integration test with lesser details.
For example there could be domain service having slipped into domain, a domain service using an application service or a repository, etc. Just a hand full of violations should be fine...
Do you want to add such an example (instead of the current one), or should I do that?

Reply:
I like this idea and prefer that you add the example. I think it is better to have an external view on the topic and provide an example that is different from my first suggestions. This should also speed up the PR as I really want to use the feature in production :)
Concerning the documentation: Is it sufficient? Can you do a short review on it as well? I think the real world example might also be handy for the documentation. As stated above, I fail building the documentation locally. Any hints for that?

Reply:
Okay, I'll try to add an example and let you review it :smiley: 
About the docs, wasn't your problem that you haven't executed `./gradlew` within the docs folder to regenerate the user guide? You need to check in the generated html. This would be a lot easier if GitHub Pages would whitelist the Jekyll Asciidoc plugin :frowning_face: